### Reorganisation_Aug2014
## HPV HNSC APOBEC
This project has given rise to a recent paper Cell Reports 2014, and a short additional report in Mol. Cell Onc. 2014. There should be a grant written on this soonish:

1. Various derived bits of data are on my Mac desktop in HPV_HNSC folder. Further details are in a README file but briefly it includes CSV files with:
- The expression of E6 and E7 for all HPV samples
- Clinical data for all HNSC, 
- mutation counts (neutral, TTW, TGW) for all HNSC samples, 
- the same plus for CESC samples. 
- The estimates of HPV copy no derived from HPV coverage/WGS coverage.
- R scripts for ternary plots used in MCO paper.
- RData of full RNAseq data for all HPV samples.
- A cryptic table from Gad Getz retrotransposon paper that I dont fully understand, maybe need to mail them about.
2. Further data can be found on Legion (/home/rmgzshd/Scratch/HPV_Proj)
- The hpv.bam files were created by realigning non paired reads from TCGA HPV+ HNSC bam files to the HPV genome (/home/rmgzshd/Scratch/HPV_Proj) using bowtie2. The bash scripts (HPVcoverage.sh, HG19coverage.sh) for this are on my mac Git folder (and Github) 
- The .integration.bed files contain sites where an alignment pair has one foot in HG19 genomic sequence and one in HPV genome (same scripts as above).
3. The heavy lifting was actually done on my CI server, and the list of TCGA HNSC samples that were analysed can be found in /mnt/store1/cghub/HNSC/SampleData within the manifest .xml files.

## MISO STAR NF2 RNAseq
I had previously done some RNAseq analysis for Pablo Viciana-Rodriguez on NF2 mutant cell lines. I was happy with the transcript level expression eestimation for this (Sailfish alignment, limma voom normalisation and statistical difference). I was not happy with the exon level analysis (Tophat, Cuffdiff, Cufflinks etc). A colleague of Javier has recommended MISO and STAR and both have been installed on the Legion server.
1. **Genomes**. After a few false starts I am using the UCSC hg19 genomic data downloaded from http://support.illumina.com/sequencing/sequencing_software/igenome.ilmn for STAR and MISO (/home/rmgzshd/Scratch/Homo_sapiens). The STAR index created from this genome is in Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/. Known alternative splice events were downloaded from the MISO site and are in /home/rmgzshd/Scratch/Homo_sapiens/UCSC/hg19/Annotation/MISO_hg19_altevents. I will try where possible to use the UCSC build for other projects too. 
2. **Code**. The scripts for STAR and MISO are in the Git folder /Users/rmgzshd/Bash/Legion. The steps are in order:
- **STAR_build.sh** used once to create the STAR alignment index from the UCSC whole genome fasta files (~2hrs).
- **STAR.sh** runs STAR alignment to create BAM files (and Log and SJ splice junction files) from each of the NF2 FASTQ files (~20mins per FASTQ).
- **MISOindexing.sh** This creates an index of splice events in the **indexed** folder (~30mins).
- **MISOutils.sh** Uses known long single variant exons (written to **exons** folder) to estimate the insert length from the data. This is used as a parameter in subsequent steps/scripts (~30mins)
- **MISO.sh** Here there is a real problem. This part of the analysis will not run on SGE (ie the Legion cluster) as advertised. Essentially the first step will do some initial calculations then cretae a whole set of autogenerated scripts from the data that are then supposed to split the job init bits and send to cluster. But you cannot send to cluster from a job already running on the cluster (AFAIK). SO this first bit basically runs on the head node (takes about an hour). I then ***qsub -l h_rt=01:00:00 -l mem=32G -l tmpfs=32G -P TCGAHPVHNSC -N MISO -wd /home/rmgzshd/Scratch/Pablo/MISO/SC1_ATCACG_L007/cluster_scripts /home/rmgzshd/Scratch/Pablo/MISO/SC1_ATCACG_L007/cluster_scripts/run_miso.sh*** it of to the cluster myself. Not really a pipeline, not automated, not easy to keep track of, not nice. (~ 20 mins per sample on head node, and 1hr30mins per sample qsub)
- **MISOcompare** At this point the results of the qsub job above are compared for differences in splice variants between samples... I have not fully tried this yet. (~dont know mins)
3. **MISO Organisation**. The MISO software itself works with a set folder structure. The MISO folder is at /home/rmgzshd/Scratch/Pablo/MISO and contains:
- **exons** output from **MISOutils** above. 
- **indexed** output from **MISOindexing** above. 
- **insert-dist** output from **MISOutils** above 
- **settings** the settings file is a parameter in the MISO.sh main script but as its main function seems to be to set qsub parameters that wont work it is probably redundant. however the main script won't run unless it is extant and correctly formulated.

Not sure whether to try and rewrite bits of MISO to run on Legion properly. I am decent with Python coding but know fuck all about Legion or Python setup.

## MRAS SHOC2 Expression
Here Pablo has asked me to look at the expression of SHOC2 and MRAS in KRAS mutant tumours as he has some in-vitro data that suggests a link. So I have set out to compare the genetic data on KRAS mutations available from Tumorportal: http://cancergenome.broadinstitute.org/index.php with expression data downloaded from TCGA: https://tcga-data.nci.nih.gov/tcga/dataAccessMatrix.htm. The available tumours with the most KRAS mutations in the TCGA were UCEC, CRC, LUAD and MM - or endometrial, colorectal, lung adenocarcinoma, and multiple myeloma. After downloading and processing RNAseq data I surprisingly found that the genetic and RNAseq data only substantially overlapped for UCEC (n=241) and LUAD (n=171).

1. **Data and Code** This is on my Mac desktop in TCGA_RNAseq folder. The processing of individual sample RNAseq into a single expression matrix - and filtering to only those also sequenced in Tumorportal is in the file_munging.R script.
2. **Results** The expression of SHOC2 and MRAS in KRAS mutatnt and WT tumours are in SHOC2_MRASexprs.tiff. Still some follow-up to de done...

Pablo seems pleased with results - couple more things to do this week.

## CAKE
CAKE is actually a framework for several genome variant callers: mpileup,varscan,bambino,somaticsniper. These all give slightly different results with no consensus on which is best. Most people seem to agree that the overlap or agreement from a few of these is then a good approach to variant calling. I am using this for the HPV project - in particular to call stretched of point mutations (kataegesis) in HNSC and maybe CESC samples. But I think this will be a useful tool for BLIC projects to come. It is installed on Legion.

1. **Genomes**. I aim to use the UCSC hg19 iGenomes (as for STAR and MISO). This is what I have used previously for analyses on HPV HNSC above. Most of the necessary info is specified in a settings file SomaticCallerConfig.pm (in the Legion Git folder). So far the genome annotations include:
- whole genome fasta (indexed using samtools) /home/rmgzshd/Scratch/Homo_sapiens/UCSC/hg19/Sequence/WholeGenomeFasta/genome.fa
- known SNPs taken from the 1000 genomes /home/rmgzshd/Scratch/Homo_sapiens/1000genomes/ALL.wgs.phase1_release_v3.20101123.snps_indels_sv.sites.pos
- Known exons for filtering variants from exome sequencing derived from biomart (all exons with an associated UCSC ID) /home/rmgzshd/Scratch/Homo_sapiens/UCSC/hg19/Annotation/Exons/UCSCexons.txt
- Known indel regions again from the 1000 genome  were taken from a VCF file and converted to a simpler format using bash, ***tail -n 14 ALL.phase1_exome_indel_consensus_regenotyped_withAA.20110521.indels.exome.genotypes.vcf |awk '{print $1,$2,(length($4)>length($5))?length($4)+$2-1:length($5)+$2-1}' > indels.txt*** . The link is /home/rmgzshd/Scratch/Homo_sapiens/UCSC/hg19/Annotation/Indels/indels.txt
- more stuff to follow
2. **No CAKE scripts yet**






















